### PR TITLE
74284 business logic for steps suppliers preservation

### PR DESCRIPTION
**Pull request by llykkelarsen**
add business logic:
- supplier step (step which has mode w/ 'for supplier') may only be first in the list of steps (though the list may also be without any supplier step)
- there may only be one supplier step for each journey
- supplier step may not be swapped 

tests

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/IStepValidator.cs [hunk 1]
```diff
@@ -9,5 +9,6 @@ public interface IStepValidator
         Task<bool> ExistsAsync(int journeyId, string stepTitle, CancellationToken token);
         Task<bool> ExistsInExistingJourneyAsync(int stepId, string stepTitle, CancellationToken token);
         Task<bool> IsVoidedAsync(int stepId, CancellationToken token);
+        Task<bool> IsSupplierStep(int stepAId, int stepBId, CancellationToken token);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/IStepValidator.cs:12]: Rename to IsAnySupplierStep ?

File: src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs [hunk 2]
```diff
@@ -40,5 +41,11 @@ public async Task<bool> IsVoidedAsync(int stepId, CancellationToken token)
                 select s).SingleOrDefaultAsync(token);
             return step != null && step.IsVoided;
         }
+
+        public async Task<bool> IsSupplierStep(int stepAId, int stepBId, CancellationToken token) 
+            => await (from s in _context.QuerySet<Step>()
+                join mode in _context.QuerySet<Mode>() on s.ModeId equals EF.Property<int>(mode, "Id")
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs:47]: change EF.Property<int>(mode, "Id") to mode.Id

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs [hunk 3]
```diff
@@ -15,20 +15,31 @@ public class StepValidatorTests : ReadOnlyTestsBase
...
         private const string StepTitle2InJourney1 = "Step3";
+        private const string StepTitle3InJourney1 = "Step4";
+        private const string StepTitle4InJourney1 = "Step5";
 
         protected override void SetupNewDatabase(DbContextOptions<PreservationContext> dbContextOptions)
         {
             using (var context = new PreservationContext(dbContextOptions, _plantProvider, _eventDispatcher, _currentUserProvider))
             {
-                var mode = AddMode(context, "M");
+                var mode = AddMode(context, "M", true);
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs:31]: rename to supplierMode

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs [hunk 4]
```diff
@@ -15,20 +15,31 @@ public class StepValidatorTests : ReadOnlyTestsBase
...
                 var responsible = AddResponsible(context, "R");
 
-                _journey1 = AddJourneyWithStep(context, "J1", StepTitle1InJourney1, mode, responsible);
-                _step1InJourney1 = _journey1.Steps.Single();
-                _journey1.AddStep(new Step(TestPlant, StepTitle2InJourney1, AddMode(context, "M2"), AddResponsible(context, "R2")));
+                _journey1 = AddJourneyWithStep(context, "J1", StepTitle2InJourney1, mode, responsible);
+                _stepForSupplierInJourney1 = _journey1.Steps.Single(s => s.Title == StepTitle2InJourney1);
+                _step2ForSupplierInJourney1 = new Step(TestPlant, StepTitle4InJourney1, AddMode(context, "M1", true), AddResponsible(context, "R1"));
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs:36]: Ehh .. don't understand .. isn't the result here that you have 2 steps in journey and both have ForSupplier in Mode?

Comment (llykkelarsen) [reply src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs:36]: Yes.. For the UpdatingNotFirstStepToSupplierStep test I needed to have a supplier step that was not first in the list

File: src/Equinor.Procosys.Preservation.Command/JourneyCommands/CreateStep/CreateStepCommandValidator.cs [hunk 5]
```diff
@@ -49,6 +51,9 @@ async Task<bool> NotBeAVoidedJourney(int journeyId, CancellationToken token)
 
             async Task<bool> NotBeAVoidedModeAsync(int modeId, CancellationToken token)
                 => !await modeValidator.IsVoidedAsync(modeId, token);
+
+            async Task<bool> BeFirstIfSupplierStep(int journeyId, int modeId, CancellationToken token)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/JourneyCommands/CreateStep/CreateStepCommandValidator.cs:55]: BeFirstStepIfModeIsForSupplier

File: src/Equinor.Procosys.Preservation.Command/JourneyCommands/CreateStep/CreateStepCommandValidator.cs [hunk 6]
```diff
@@ -49,6 +51,9 @@ async Task<bool> NotBeAVoidedJourney(int journeyId, CancellationToken token)
...
+                => await journeyValidator.IsFirstStepInAJourneyIfASupplierStepAsync(journeyId, modeId, token);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/JourneyCommands/CreateStep/CreateStepCommandValidator.cs:56]: IsFirstStepIfModeIsForSupplier

File: src/Equinor.Procosys.Preservation.Command/JourneyCommands/SwapSteps/SwapStepsCommandValidator.cs [hunk 7]
```diff
@@ -32,6 +34,8 @@ async Task<bool> BeAnExistingStepBAsync(int stepId, CancellationToken token)
                 => await stepValidator.ExistsAsync(stepId, token);
             async Task<bool> BeAdjacentStepsInAJourneyAsync(int journeyId, int stepAId, int stepBId, CancellationToken token)
                 => await journeyValidator.AreAdjacentStepsInAJourneyAsync(journeyId, stepAId, stepBId, token);
+            async Task<bool> NotIncludeSupplierStep(int stepAId, int stepBId, CancellationToken token)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/JourneyCommands/SwapSteps/SwapStepsCommandValidator.cs:37]: NotIncludeAnySupplierStep

File: src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs [hunk 8]
```diff
@@ -40,5 +41,28 @@ public async Task<bool> IsVoidedAsync(int stepId, CancellationToken token)
...
             return step != null && step.IsVoided;
         }
+
+        public async Task<bool> IsAnySupplierStep(int stepAId, int stepBId, CancellationToken token)
+            => await (from s in _context.QuerySet<Step>()
+                join mode in _context.QuerySet<Mode>() on s.ModeId equals mode.Id
+                where (s.Id == stepAId || s.Id == stepBId) && mode.ForSupplier
+                select mode).AnyAsync(token);
+
+        public async Task<bool> IsAllowedToUpdate(int journeyId, int modeId,
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs:51]: This name confuse me
What about BeFirstStepOrModeNotForSupplier?

File: src/Equinor.Procosys.Preservation.Command/JourneyCommands/SwapSteps/SwapStepsCommandValidator.cs [hunk 9]
```diff
@@ -32,6 +34,8 @@ async Task<bool> BeAnExistingStepBAsync(int stepId, CancellationToken token)
...
+                => !await stepValidator.IsAnySupplierStep(stepAId, stepBId, token);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/JourneyCommands/SwapSteps/SwapStepsCommandValidator.cs:38]: IsAnyStepForSupplier

File: src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs (hunk 8 above)
Comment (llykkelarsen) [reply src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs:51]: I was unsure about the wording of most of these methods. IsFirstStepOrModeIsNotForSupplier is much better.

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/JourneyValidatorTests.cs [hunk 10]
```diff
@@ -13,19 +14,29 @@ public class JourneyValidatorTests : ReadOnlyTestsBase
     {
         private const string JourneyTitle = "Journey";
         private const string JourneyTitle2 = "Journey2";
+        private const string JourneyTitle3 = "Journey3";
+        private const string ModeTitle1 = "M";
+        private const string ModeTitle2 = "M2";
 
-        private int _journeyId;
-        private int _journeyId2;
+        private int _journeyWithStepId;
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/JourneyValidatorTests.cs:21]: _journeyWithoutSupplierStepId

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs (hunk 4 above)
Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs:36]: In my head you should create that test by having a journey with 2 non Supplier-Steps.
A positive test is making the first one to be Supplier Step
A negative test is trying to do the same with step 2

**Comment by llykkelarsen** [comment]
renaming is ready for review

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs (hunk 4 above)
Comment (llykkelarsen) [reply src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs]: I see what you're saying now. Fixed in my most recent commit

**Pull request merged**